### PR TITLE
Fix long timestamps conversion

### PR DESCRIPTION
**Pull request by fevac**
## Description

Closes https://github.com/Clinical-Genomics/trailblazer/issues/284

### Fixed

- Timestamp conversion with long decimal numbers

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b fix_long_timestamps -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: trailblazer/utils/datetime.py [hunk 1]
```diff
@@ -38,6 +38,9 @@ def get_datetime_from_timestamp(timestamp: str, datetime_formats: List[str]) ->
 def tower_datetime_converter(timestamp: str) -> datetime:
     """Converts a NF Tower timestamp into a datetime object."""
     allowed_formats = [TOWER_TIMESTAMP_FORMAT, TOWER_TIMESTAMP_FORMAT_EXTENDED]
+    # old python versions do not handle high precision numbers. Errors can be avoided by capping to 26 characters
+    if len(timestamp) > 25:
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply trailblazer/utils/datetime.py:42]: maybe `MAX_NUMBER_OF_DIGITS` and delete the inline comment

File: trailblazer/utils/datetime.py [hunk 2]
```diff
@@ -38,6 +38,9 @@ def get_datetime_from_timestamp(timestamp: str, datetime_formats: List[str]) ->
...
+        timestamp = f"{timestamp[:25]}Z"
```
Comment (diitaz93) [inline trailblazer/utils/datetime.py:43]:
```suggestion
        timestamp: str = f"{timestamp[:25]}Z"
```

File: trailblazer/utils/datetime.py [hunk 3]
```diff
@@ -38,6 +38,9 @@ def get_datetime_from_timestamp(timestamp: str, datetime_formats: List[str]) ->
...
     return get_datetime_from_timestamp(timestamp, allowed_formats)
```
Comment (diitaz93) [inline trailblazer/utils/datetime.py:44]: use named args as the function has 2+ arguments

File: trailblazer/utils/datetime.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (diitaz93) [reply trailblazer/utils/datetime.py:42]: I think your tests are failing because the last 2 decimals don't match. Maybe reducing to 23 could fix that

**Review by diitaz93 — COMMENTED** [review]
Looks good, lets see if with 23 decimals the tests pass

File: trailblazer/utils/datetime.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fevac) [reply trailblazer/utils/datetime.py]: fixed with 26 digits

File: tests/utils/test_datetime.py [hunk 4]
```diff
@@ -87,3 +88,19 @@ def test_get_date_number_of_days_ago(timestamp_now: datetime):
...
+    "timestamp",
+    ["2023-04-04T08:11:24Z", "2023-06-20T08:01:57.661819Z", "2023-09-14T11:14:55.664772403Z"],
+)
+def test_tower_datetime_converter(timestamp: str):
+    """Test parsing a Tower timestamp."""
+
+    # GIVEN timestamps
+
+    # WHEN calling the function
+    date: datetime = tower_datetime_converter(timestamp=timestamp)
```
[the remark this answers is not in the document]
Comment (fevac) [reply tests/utils/test_datetime.py]: it is a date time stamp. Changed it now

**Comment by fevac** [comment]
@henrikstranneheim @diitaz93 thanks for the comments. Can any of you approve it?

File: tests/utils/test_datetime.py [hunk 5]
```diff
@@ -87,3 +88,19 @@ def test_get_date_number_of_days_ago(timestamp_now: datetime):
...
     # Then the date returned should be less than then today
     assert date < timestamp_now
+
+
+@pytest.mark.parametrize(
+    "datetime_stamp",
+    ["2023-04-04T08:11:24Z", "2023-06-20T08:01:57.661819Z", "2023-09-14T11:14:55.664772403Z"],
+)
+def test_tower_datetime_converter(datetime_stamp: str):
+    """Test parsing a Tower timestamp."""
```
Comment (henrikstranneheim) [inline tests/utils/test_datetime.py:98]:
```suggestion
    """Test parsing a Tower datetime stamp."""
```

File: tests/utils/test_datetime.py [hunk 6]
```diff
@@ -87,3 +88,19 @@ def test_get_date_number_of_days_ago(timestamp_now: datetime):
...
+
+    # GIVEN timestamps
```
Comment (henrikstranneheim) [inline tests/utils/test_datetime.py:100]:
```suggestion
    # GIVE Ndatetime stamps
```

File: tests/utils/test_datetime.py [hunk 7]
```diff
@@ -87,3 +88,19 @@ def test_get_date_number_of_days_ago(timestamp_now: datetime):
...
+
+    # WHEN calling the function
+    date: datetime = tower_datetime_converter(timestamp=datetime_stamp)
```
Comment (henrikstranneheim) [inline tests/utils/test_datetime.py:103]:
```suggestion
    date: datetime = tower_datetime_converter(datetime_stamp=datetime_stamp)
```

File: trailblazer/utils/datetime.py [hunk 8]
```diff
@@ -38,7 +39,10 @@ def get_datetime_from_timestamp(timestamp: str, datetime_formats: List[str]) ->
 def tower_datetime_converter(timestamp: str) -> datetime:
```
Comment (henrikstranneheim) [inline trailblazer/utils/datetime.py:39]:
```suggestion
def tower_datetime_converter(datetime_stamp: str) -> datetime:
```

**Review by henrikstranneheim — APPROVED** [review]
Nice - just some final fixes 👍

**Review by diitaz93 — APPROVED** [review]
Looks good 👍

**Comment by fevac** [comment]
Tests cannot be performed as long datetime stamps cannot be predicted. But pytests were added and `trailblazer scan` is working

**Pull request merged**